infosec-orverview.md

```diff
@@ -0,0 +1,29 @@
+* All connections are initiated by the _collaborator_.
+* The _collaborator_ does not open any listening sockets.
+* Connections are secured using mutually-authenticated TLS.
+* Each request response pair is done on a new TLS connection.
+* The PKI for OpenFL federations can be created using openssl tools, with the organization hosting the _aggregator_ usually acting as the Certificate Authority (CA) that verifies each identity before signing.
+* Currently, the _collaborator_ polls the _aggregator_ at a fixed interval. We have had a request to enable client-side configuration of this interval and hope to support that feature soon.
+* Connection timeouts are set to gRPC defaults.
+* If the _aggregator_ is not available, the _collaborator_ will retry connections indefinitely. This is currently useful so that we can take the aggregator down for bugfixes without _collaborator_ processes exiting.
+
+#### Overview of Contents of Network Messages
+Network messages are well defined protobufs which can be found in [aggregator.proto](https://github.com/securefederatedai/openfl/blob/develop/openfl/protocols/aggregator.proto) and [base.proto](https://github.com/securefederatedai/openfl/blob/develop/openfl/protocols/base.proto).
+Key points about the network messages/protocol:
+* No private data is ever sent to the _aggregator_
+* No executable code is ever sent to the _collaborator_. All algorithms come with the OpenFL package and the custom FL workspace.
+* The _collaborator_ typically requests the FL tasks to execute from the aggregator via the `GetTasksRequest` message
+* The _aggregator_ reads the FL plan and returns a `GetTasksResponse` which includes metadata (`Tasks`) about the Python functions to be invoked by the collaborator (the code being installed locally as part of a pre-distributed workspace bundle)
+* The _collaborator_ then uses its TaskRunner framework to execute the FL tasks on the locally available data, producing output tensors such as model weights or metrics
+* During task execution, the _collaborator_ may additionally request tensors from the aggregator via the `GetAggregatedTensor` RPC method
+* Upon task completion, the _collaborator_ transmits the results by emitting a `SendLocalTaskResults` call which contains `NamedTensor` objects that encode model weight updates or ML metrics such as loss or accuracy (among others).
+
+#### Testing a Collaborator
+In order to test a _collaborator_, the _aggregator_ can be set to not aggregate updates from specific _collaborators_ so that they can test functionality without impacting the running model training/validation. We hope this enables InfoSec runtime analysis against the live _aggregator_ without the need to access any private data on the _collaborator_ machine running the test. To do this, please coordinate with the OpenFL aggregator admins.
```

Comment:
As of [the April 15th OpenFL nightly build](https://pypi.org/project/openfl-nightly/1.9.0.dev20250415/), we support `no-op` workspace, which can be used in conjunction with the recently introduced `fx collaborator ping` command to test collaborator/aggregator connectivity without running any FL tasks or accessing private data.

PS: this functionality will be officially available in the upcoming OpenFL 1.9 release

Reply:
A few thoughts:

Maybe " No private data is ever sent to the _aggregator_" should be removed, as this is not guaranteed by the messaging protocol itself. I could directly embed the data into a model sized tensor and send as a "model update".

I would expand the next bullet to read *something like*, "No executable code is ever sent to the _collaborator_. All code to be executed resides in the combination of the OpenFL package and custom FL workspace. These codes, along with the FL plan file that defines the classes and initial parameters to be used, is available for review before the federation runs in order that all possible execution is understood before it occurs."